File: algorithm/leetcode/classical150item/hash/isIsomorphic.py
class ExampleIsIsomorphic:
    def __init__(self, s: str, t: str, answer: bool):
        self.s = s
        self.t = t
        self.answer = answer

def is_isomorphic(s: str, t: str) -> bool:
    mapping_s_to_t = {}
    mapping_t_to_s = {}

    for char_s, char_t in zip(s, t):
        if (char_s in mapping_s_to_t and mapping_s_to_t[char_s] != char_t) or \
                (char_t in mapping_t_to_s and mapping_t_to_s[char_t] != char_s):
            return False
        mapping_s_to_t[char_s] = char_t
        mapping_t_to_s[char_t] = char_s

    return True

if __name__ == "__main__":
    examples = [
        ExampleIsIsomorphic("egg", "add", True),
        ExampleIsIsomorphic("badc", "baba", False),
    ]

    for example in examples:
        result = is_isomorphic(example.s, example.t)
        print(f"{result}\n{example.answer}\n---------------")

File: algorithm/leetcode/classical150item/hash/isIsomorphic.go
package main

import "fmt"

type exampleIsIsomorphic struct {
	s      string
	t      string
	answer bool
}

func main() {
	s := []exampleIsIsomorphic{
		{"egg", "add", true},
		{"badc", "baba", false},
	}
	for _, v := range s {
		fmt.Printf("%v\n%v\n---------------\n", isIsomorphic(v.s, v.t), v.answer)
	}
}

func isIsomorphic(s string, t string) bool {
	ms, mt := make(map[byte]byte), make(map[byte]byte)
	for i := range s {
		x, y := s[i], t[i]
		if ms[x] != 0 && ms[x] != y || mt[y] != 0 && mt[y] != x {
			return false
		}
		ms[x], mt[y] = y, x
	}
	return true
}
